distributions/openhab/src/main/resources/bin/update.ps1

```diff
@@ -69,190 +12,582 @@ Function Update-openHAB {
-            $DL = DownloadFiles $DownloadLocation "$TempDir\openhab-$OHVersion.zip" $SkipNew
-            if ($DL -eq 2) {Return}
-        } else {
-            $DownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab%2F$OHVersion%2Fopenhab-$OHVersion.zip"
-            $AddonsDownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab-addons%2F$OHVersion%2Fopenhab-addons-$OHVersion.kar"
-            $LegacyAddonsDownloadLocation="https://bintray.com/openhab/mvn/download_file?file_path=org%2Fopenhab%2Fdistro%2Fopenhab-addons-legacy%2F$OHVersion%2Fopenhab-addons-legacy-$OHVersion.kar"
-            Write-Host -ForegroundColor Cyan "Downloading the openHAB $OHVersion distribution..."
-            $DL = DownloadFiles $DownloadLocation "$TempDir\openhab-$OHVersion.zip" $SkipNew
-            if ($DL -eq 2) {Return}
+        ########### STEP 4 - Run new update.ps1 if needed
+        $newUpdate = Join-Path $TempDistribution "/runtime/bin/update.ps1"
+        if ($KeepUpdateScript) {
+            try {
+                Copy-Item -Path "$OHDirectory/runtime/bin/common.psm1" -Destination "$TempDistribution/runtime/bin/common.psm1" -Force
+                Copy-Item -Path "$OHDirectory/runtime/bin/update.ps1" -Destination $newUpdate -Force
+            } catch {
+                Write-Host -ForegroundColor Magenta "Could not copy the update.ps1 to the new distribution"
+                # Don't bother with AutoConfirm here since this is special debugging logic to begin with
+                $confirmation = Read-Host "Okay to Continue? [y/N]"
+                if ($confirmation -ne 'y') {
+                    return PrintAndReturn "Cancelling update"
+                }
+            }
+        }
+        
+        If (Test-Path $newUpdate) {
+            Write-Host ""
+            BoxMessage "New update.ps1 was found - executing it instead (found in $newUpdate)" Magenta
+            Write-Host ""
+            try {
+                $rawOHVersion = $OHVersion
+                if ($rawOHVersion.EndsWith("-SNAPSHOT")) {
+                    $rawOHVersion = $rawOHVersion.Substring(0, $rawOHVersion.Length - "-SNAPSHOT".Length)
+                }
+                # go back to our original directory so the new update script does it as well
+                Set-Location -Path $StartDir  -ErrorAction Continue 
+                . $newUpdate
+                Update-openHAB -OHDirectory $OHDirectory -OHVersion $rawOHVersion -Snapshot $Snapshot -AutoConfirm $AutoConfirm -SkipNew $true -KeepUpdateScript $KeepUpdateScript
+                return 2;
+            } catch {
+                return PrintAndReturn "Execution of new update.ps1 failed - please execute it yourself (found in $newUpdate)" $_
+            }
         }
+    }
 
+    # Backup openHAB only if not coming via new update script
```

Comment:
I thought I was going crazy... This is all well and good....
...but you've forgotten to download the files! :-D

Reply:
Really not sure what you mean here about "you've forgotten to download the files"?  

If I *think* what you mean (which probably a bad assumption!), the files were downloaded, we found a new update script, executed the update script and use the files that were already downloaded.  So what am I missing?

Reply:
Nope. If you decide to `$SKIPNEW -eq $True`, you have never downloaded the new files. The only place they are downloaded is within the `if ($SkipNew -eq $False)` code. You handle the Addons and LegacyAddons but not the actual distro.